### ISSUE TITLE
Cannot create class io.flutter.project.FlutterIconProvider from IDEA/2020.3 EAP

### ISSUE THREAD

**Issue by devoncarew** [flutter/flutter-intellij#4937]
We're seeing these issues from IntelliJ IDEA/2020.3 EAP:

```
Cannot create class io.flutter.project.FlutterIconProvider [Plugin: io.flutter]
	at c.i.serviceContainer.ComponentManagerImpl.instantiateClass(ComponentManagerImpl.kt:653)
	at c.i.o.extensions.impl.ExtensionComponentAdapter.instantiateClass(ExtensionComponentAdapter.java:50)
	at c.i.o.extensions.impl.XmlExtensionAdapter$SimpleConstructorInjectionAdapter.instantiateClass(XmlExtensionAdapter.java:133)
	at c.i.o.extensions.impl.ExtensionComponentAdapter.createInstance(ExtensionC...
```

**Comment by devoncarew** [comment]
From looking at this, it's not clear to me why the `FlutterIconProvider` can't be instantiated in 2020.3 EAP.

**Comment by devoncarew** [comment]
The https://github.com/JetBrains/intellij-community/blob/master/platform/core-api/src/com/intellij/ide/IconProvider.java abstract class hasn't changed in quite a while ...

**Comment by beansoft** [comment]
Caused by: java.lang.NoSuchMethodError: 'javax.swing.Icon com.intellij.psi.impl.ElementBase.overlayIcons(javax.swing.Icon[])'
	at io.flutter.project.FlutterIconProvider.<clinit>(FlutterIconProvider.java:32)

no error on 2020.3 EAP 1, but EAP 2 failed to load.

**Comment by beansoft** [comment]
I've found the reason, the Jetbrains' developer had removed this method overlayIcons(javax.swing.Icon[]) from latest IJ CE code in ElementBase.java, but the LayeredIcon.java stays untouched.
The git log said:
platform.core.impl — remove dependency on intellij.platform.util.ui and intellij.platform.core.ui
java.psi.iml — remove dependency on intellij.platform.core.ui

My fix is create a util method it:

```java
// import static com.intellij.psi.impl.ElementBase.overlayIcons;
  @NotNull
  public static Icon overlayIcons(Icon @NotNull ... icons) {
    final LayeredIcon icon = new LayeredIcon(icons.length);
    int i = 0;
    for (Icon ic : icons) {
      icon.setIcon(ic, i++);
    }
    return icon;
  }
```

**Comment by devoncarew** [comment]
(see also https://github.com/flutter/flutter-intellij/issues/4939)

**Comment by devoncarew** [comment]
> I've found the reason, the Jetbrains' developer had removed this method overlayIcons(javax.swing.Icon[]) from latest IJ CE code in ElementBase.java, but the LayeredIcon.java stays untouched.

Ah, from https://github.com/flutter/flutter-intellij/blob/master/src/io/flutter/project/FlutterIconProvider.java#L32 - great catch.

> My fix is create a util method:
> 
> ```java
> // import static com.intellij.psi.impl.ElementBase.overlayIcons;
>   @NotNull
>   public static Icon overlayIcons(Icon @NotNull ... icons) {
>     final LayeredIcon icon = new LayeredIcon(icons.length);
>     int i = 0;
>     for (Icon ic : icons) {
>       icon.setIcon(ic, i++);
>     }
>     return icon;
>   }
> ```

Sounds good to me. I'm curious how our build system missed this - we build against the platforms and versions we're targeting.

**Comment by devoncarew** [comment]
Ah:

> no error on 2020.3 EAP 1, but EAP 2 failed to load.

Sounds like we might be building against a just slightly older version of EAP.

**Comment by beansoft** [comment]
> Ah:
> 
> > no error on 2020.3 EAP 1, but EAP 2 failed to load.
> 
> Sounds like we might be building against a just slightly older version of EAP.

Yes, they just released EAP 2 at Sep 30, 2020.3 EAP build 203.4203.26, quite a new build.
https://www.jetbrains.com/idea/nextversion/#section=mac
